2507 Added appellate courts to RSS scraper

albertisfu opened the pull request:
As required in #2507 this PR enables support to scrape appellate RSS Feeds.

Some changes were needed:

- Since appellate RSS feeds don't contain a `pacer_case_id`  we'll add dockets without this field, so it was needed to tweak the `find_docket_object` method so we don't look up dockets by blank case id otherwise it will return a bunch of wrong matches. Then when another source like the extension updates an appellate case it will be identified by the `docket_number_core` and the `pacer_case_id` will be added.

- Using the `dateparser` to parse the RSS `last_build_date_b` since for appellate it uses a local timezone code.

- Enabled appellate courts in `scrape_rss` using `all_pacer_courts()` instead of `district_pacer_courts()`

- Updated the coverage page with appellate courts.

Some tasks to do:

- If this seems right, before it's merged I'll do a Juriscraper release and update the version here.
- Once this is merged we'll need to set the `pacer_has_rss_feed` field to True for the appellate courts that have a RSS Feed, I think we could do this manually since they're only a few of them, right?

- You mention:
> Of course we need to update our daemon to actually do the scraping/merging.

Do I need to do something additional here to get the daemon updated? or it's just a deployment task?

albertisfu commented:
Working on the `Troller BK` project, I realized that there is a case where a docket can be duplicated even though we have one for that case.

The problem happens if the existing docket has already a `pacer_case_id`, then when adding merging an appellate docket where we don't have a `pacer_case_id` the lookup is made by `pacer_case_id=None` and `docket_number_core`, so the existing docket with `pacer_case_id` is not found. 

In order to solve it, I added an additional lookup using only the `docket_number_core` so it can find the existing docket with the same `docket_number_core` instead of creating a new one.

mlissner reviewed (changes requested):
This looks pretty good, but I worry about the changes to the merger code. Thank you.

albertisfu commented on cl/simple_pages/templates/help/alert_help.html:
```diff
@@ -127,6 +127,7 @@ <h4 class="v-offset-above-2">Full RSS Feeds</h4>
     <ul class="nav nav-tabs" role="tablist">
       <li role="presentation" class="active"><a href="#full-feeds-FD" aria-controls="full-feeds-district" role="tab" data-toggle="tab">District Courts</a></li>
       <li role="presentation"><a href="#full-feeds-FB" aria-controls="full-feeds-bankr" role="tab" data-toggle="tab">Bankruptcy Courts</a></li>
+      <li role="presentation"><a href="#full-feeds-F" aria-controls="full-feeds-appellate" role="tab" data-toggle="tab">Appellate Courts</a></li>
```
Done!

albertisfu commented on cl/simple_pages/templates/help/alert_help.html:
```diff
@@ -181,6 +183,7 @@ <h4 class="v-offset-above-2">No RSS Feeds</h4>
     <ul class="nav nav-tabs" role="tablist">
       <li role="presentation" class="active"><a href="#no-feeds-FD" aria-controls="no-feeds-district" role="tab" data-toggle="tab">District Courts</a></li>
       <li role="presentation"><a href="#no-feeds-FB" aria-controls="no-feeds-bankr" role="tab" data-toggle="tab">Bankruptcy Courts</a></li>
+      <li role="presentation"><a href="#no-feeds-F" aria-controls="no-feeds-appellate" role="tab" data-toggle="tab">Appellate Courts</a></li>
```
Sure, changed.
I also tweaked the title from:
`In addition to the federal Circuit Courts, the following courts do not provide RSS feeds of their latest filings:`

To:
`The following courts do not provide RSS feeds of their latest filings:`

albertisfu commented on poetry.lock:
```diff
@@ -41,14 +41,14 @@ files = [
 [[package]]
 name = "asgiref"
-version = "3.5.0"
+version = "3.6.0"
```
Yeah, I've fixed the poetry lock, now it only includes the new dependencies.

albertisfu commented on cl/recap/mergers.py:
```diff
@@ -80,20 +80,27 @@ def find_docket_object(
         )
+        lookups.append(
+            {"docket_number_core": docket_number_core},
+        )
```
Yeah about this questions:

> Why do we need this lookup? Earlier we match on pacer_case_id + docket_number_core, then we try on no pacer_case_id + docket_number_core. I guess this would match on docket_number_core with a mismatched pacer_case_id, which seems bad, doesn't it?

Yeah, I added this new lookup because I realized a case where for instance, the appellate docket is added first by the extension and they do have a `pacer_case_id` (like the new appellate dockets), then when the scraper starts scraping appellate courts and we get data for one of those dockets that have a `pacer_case_id`, since the RSS data don't contains a `pacer_case_id` the first three lookups wont' work:

```
{"pacer_case_id": pacer_case_id, "docket_number_core": docket_number_core,}
{"pacer_case_id": pacer_case_id}
{"pacer_case_id": None, "docket_number_core": docket_number_core,}
```

Since we don't have the `pacer_case_id` in the RSS data and also the `pacer_case_id` is not `None` in the docket.
So in those cases a new docket will be created instead of updating the existing one.

That's why I added a lookup only using the `docket_number_core` so that the docket can be found.
This example is tested in: `test_appellate_merge_rss_feed_with_case_id`

But yes, as you mention this might be risky so a docket with the wrong `pacer_case_id` can be matched since we don't have the `pacer_case_id` to compare it. 
Would it help if we just apply this lookup for appellate dockets? 
In #2511 we do an additional check confirming the `docket_number` but not sure if that would be enough.

> Assuming there's a logical answer for the first question, why do these as two different appends rather than just doing one?

Changed, now using extend to add the lookups.

mlissner replied:
Got it. Hopefully this should be OK. We'll see, I guess.

albertisfu commented:
Thanks! I've applied the requested changes and answered your questions above.

Let me know what do you think.

mlissner reviewed (changes requested):
OK, code looks all good to me. We just need to get the dependencies cleaned up (you have a reference to a git commit in pyproject.toml), then this can merge.

Thanks.

albertisfu commented:
Done! I've updated the pyproject.toml using the new Juriscraper version.

I also fixed a couple of if conditions within the `find_docket_object` method after adding the new lookup dict `{"docket_number_core": docket_number_core}`. The `pacer_case_id` is not set in this dict, so when using `kwargs["pacer_case_id"]`, it threw an error. In order to fix it I changed it to` kwargs.get("pacer_case_id")`.

I think this now can be merged, after approving the requested changes, thanks!
